thingsboard_gateway/connectors/ftp/ftp_uplink_converter.py

```diff
@@ -0,0 +1,197 @@
+        if '${' in self.__config['devicePatternType'] and '}' in self.__config['devicePatternType']:
+            get_device_type_from_data = True
+        else:
+            result['deviceType'] = self.__config['devicePatternType']
+
+        try:
+            for data_type in data_types:
+                for information in self.__config[data_type]:
+                    arr = data.split(self.__config['delimiter'])
+
+                    key_index = information['key']
+                    val_index = information['value']
+
+                    if '${' in information['key'] and '}' in information['key']:
+                        key_index = config['headers'].index(re.sub(r'[^\w]', '', information['key']))
+
+                    if '${' in information['value'] and '}' in information['value']:
+                        val_index = config['headers'].index(re.sub(r'[^\w]', '', information['value']))
+
+                    result[data_types[data_type]].append({
+                        arr[key_index] if isinstance(key_index, int) else key_index:
+                            arr[val_index] if isinstance(val_index, int) else val_index
+                    })
+
+                    if get_device_name_from_data:
+                        index = config['headers'].index(re.sub(r'[^\w]', '', self.__config['devicePatternName']))
+                        result['deviceName'] = arr[index]
+                    if get_device_type_from_data:
+                        index = config['headers'].index(re.sub(r'[^\w]', '', self.__config['devicePatternType']))
+                        result['deviceType'] = arr[index]
+
+        except Exception as e:
+            log.error('Error in converter, for config: \n%s\n and message: \n%s\n', dumps(self.__config), data)
+            log.exception(e)
+
+        return result
+
+    def _convert_slices_view_data(self, data):
+        get_device_name_from_data = False
+        get_device_type_from_data = False
+
+        data_types = {'attributes': 'attributes', 'timeseries': 'telemetry'}
+        result = {
+            'deviceName': 'asd',
+            'deviceType': 'Device',
+            'attributes': [],
+            'telemetry': []
+        }
+
+        if '[' in self.__config['devicePatternName'] and ']' in self.__config['devicePatternName']:
+            get_device_name_from_data = True
+        else:
+            result['deviceName'] = self.__config['devicePatternName']
+        if '[' in self.__config['devicePatternType'] and ']' in self.__config['devicePatternType']:
+            get_device_type_from_data = True
+        else:
+            result['deviceType'] = self.__config['devicePatternType']
+
+        try:
+            for data_type in data_types:
+                for information in self.__config[data_type]:
+                    arr = data.split(self.__config['delimiter'])
+
+                    if '[' in information['value'] and ']' in information['value']:
```

Comment:
It looks like we do the same things with "key" and "value" - could we move it to some method and just a path "key"/"value" as an argument?